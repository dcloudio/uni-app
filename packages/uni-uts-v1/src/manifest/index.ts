import { existsSync, readFileSync } from 'fs-extra'
import { join } from 'path'
import { parseJson } from '../shared'
import {
  checkManifest,
  hasCustomResources,
  isCustomResources,
  resolveManifestJson,
  resolvePluginFiles,
} from './manifest'
import {
  APP_PLATFORM,
  CheckOptions,
  CheckResult,
  customResourceChangedTips,
  customResourceTips,
} from './utils'

export { genManifestFile } from './manifest'

export { restoreDex } from './dex'
export { restoreSourceMap, storeSourceMap } from './sourceMap'

interface PlatformOptions {
  customRes: string[]
}

const ANDROID_CUSTOM_RES = [
  'app-android/assets/',
  'app-android/libs/',
  'app-android/res/',
  'app-android/AndroidManifest.xml',
]

const IOS_CUSTOM_RES = [
  'app-ios/Frameworks/',
  'app-ios/Resources/',
  'app-ios/Info.plist',
]

export function checkKotlinCompile(
  playground: typeof process.env.HX_USE_BASE_TYPE,
  options: CheckOptions
) {
  return checkCompile('app-android', playground, options)
}

export function checkSwiftCompile(
  playground: typeof process.env.HX_USE_BASE_TYPE,
  options: CheckOptions
) {
  return checkCompile('app-ios', playground, options)
}

export function checkCompile(
  platform: APP_PLATFORM,
  playground: typeof process.env.HX_USE_BASE_TYPE,
  options: CheckOptions
) {
  const platformOptions: PlatformOptions = {
    customRes: platform === 'app-android' ? ANDROID_CUSTOM_RES : IOS_CUSTOM_RES,
  }
  if (playground === 'standard') {
    return checkWithPlayground(platform, 'standard', options, platformOptions)
  }
  return checkWithPlayground(platform, 'custom', options, platformOptions)
}

async function checkWithPlayground(
  platform: APP_PLATFORM,
  type: typeof process.env.HX_USE_BASE_TYPE,
  {
    id,
    env,
    cacheDir,
    pluginDir,
    pluginRelativeDir,
    is_uni_modules,
  }: CheckOptions,
  { customRes }: PlatformOptions
): Promise<CheckResult> {
  // 第一步：获取所有文件列表
  const files = await resolvePluginFiles(platform, pluginDir, is_uni_modules)
  let tips = ''
  // 标准基座检查是否包含原生资源/配置
  if (type === 'standard') {
    if (hasCustomResources(files, customRes)) {
      tips = customResourceTips(id)
    } else {
      // 检查 config.json
      if (platform === 'app-android') {
        if (androidHasCustomConfigJson(pluginDir, is_uni_modules)) {
          tips = customResourceTips(id)
        }
      } else if (platform === 'app-ios') {
        if (iOSHasCustomConfigJson(pluginDir, is_uni_modules)) {
          tips = customResourceTips(id)
        }
      }
    }
  }
  // 第二步：获取当前插件缓存文件信息
  const manifest = resolveManifestJson(platform, pluginRelativeDir, cacheDir)
  if (!manifest) {
    return { expired: true, tips, files }
  }
  // 第四步：检查文件变更
  const res = await checkManifest(manifest, { env, files, pluginDir })
  // 自定义基座检查原生资源/配置是否发生变化
  if (type === 'custom' && typeof res === 'string') {
    if (isCustomResources(res, customRes)) {
      tips = customResourceChangedTips(id)
    }
  }
  return {
    expired: res !== true,
    tips,
    files,
  }
}

export function initCheckOptionsEnv(): CheckOptions['env'] {
  return {
    compilerVersion: require('../../package.json').version,
  }
}

function androidHasCustomConfigJson(
  pluginDir: string,
  is_uni_modules: boolean
) {
  return hasCustomConfigJson(
    'app-android',
    'minSdkVersion',
    pluginDir,
    is_uni_modules
  )
}

function iOSHasCustomConfigJson(pluginDir: string, is_uni_modules: boolean) {
  return hasCustomConfigJson(
    'app-ios',
    'deploymentTarget',
    pluginDir,
    is_uni_modules
  )
}

function hasCustomConfigJson(
  platform: APP_PLATFORM,
  key: 'minSdkVersion' | 'deploymentTarget',
  pluginDir: string,
  is_uni_modules: boolean
) {
  const configJsonFile = join(
    pluginDir,
    is_uni_modules ? 'utssdk' : '',
    platform,
    'config.json'
  )
  if (configJsonFile && existsSync(configJsonFile)) {
    try {
      const configJson = parseJson(readFileSync(configJsonFile, 'utf8'))
      const len = Object.keys(configJson).length
      if (len > 1) {
        return true
      }
      if (len === 1) {
        if (!configJson.hasOwnProperty(key)) {
          return true
        }
      }
    } catch (e) {}
  }
  return false
}
