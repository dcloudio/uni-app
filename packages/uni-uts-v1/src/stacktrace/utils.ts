import fs from 'fs'
import path, { relative } from 'path'
import { normalizePath } from '../shared'

export interface GenerateRuntimeCodeFrameOptions {
  cacheDir: string
  logType?: 'log' | 'info' | 'warn' | 'debug' | 'error'
}

export const COLORS: Record<string, string> = {
  warn: '\u200B',
  error: '\u200C',
}

export const splitRE = /\r?\n/

export function resolveSourceMapDirByCacheDir(cacheDir: string) {
  return path.resolve(cacheDir, 'sourcemap')
}

export function resolveSourceMapFileBySourceFile(
  file: string,
  sourceMapDir: string
) {
  const sourceMapFile = path.resolve(sourceMapDir, file + '.map')
  if (fs.existsSync(sourceMapFile)) {
    return sourceMapFile
  }
}

const range: number = 2

function posToNumber(
  source: string,
  pos: number | { line: number; column: number }
): number {
  if (typeof pos === 'number') return pos
  const lines = source.split(splitRE)
  const { line, column } = pos
  let start = 0
  for (let i = 0; i < line - 1; i++) {
    start += lines[i].length + 1
  }
  return start + column
}

export function lineColumnToStartEnd(
  source: string,
  line: number,
  column: number
) {
  const lines = source.split(splitRE)
  let start = 0
  for (let i = 0; i < line - 1; i++) {
    start += lines[i].length + 1
  }
  return {
    start: start + column,
    end: start + lines[line - 1].length,
  }
}

export function generateCodeFrame(
  source: string,
  start: number | { line: number; column: number } = 0,
  end?: number
): string {
  start = posToNumber(source, start)
  end = end || start
  const lines = source.split(splitRE)
  let count = 0
  const res: string[] = []
  for (let i = 0; i < lines.length; i++) {
    count += lines[i].length + 1
    if (count >= start) {
      for (let j = i - range; j <= i + range || end > count; j++) {
        if (j < 0 || j >= lines.length) continue
        const line = j + 1
        res.push(
          `${line}${' '.repeat(Math.max(3 - String(line).length, 0))}|  ${
            lines[j]
          }`
        )
        const lineLength = lines[j].length
        if (j === i) {
          // push underline
          const pad = start - (count - lineLength) + 1
          const length = Math.max(
            1,
            end > count ? lineLength - pad : end - start
          )
          res.push(`   |  ` + ' '.repeat(pad) + '^'.repeat(length))
        } else if (j > i) {
          if (end > count) {
            const length = Math.max(Math.min(end - count, lineLength), 1)
            res.push(`   |  ` + '^'.repeat(length))
          }
          count += lineLength + 1
        }
      }
      break
    }
  }
  return res.join('\n')
}

export function parseRelativeSourceFile(
  sourceFile: string,
  sourceRoot?: string | null
) {
  if (!sourceRoot) {
    return sourceFile
  }
  sourceFile = normalizePath(sourceFile)
  sourceRoot = normalizePath(sourceRoot)
  if (sourceFile.startsWith(sourceRoot)) {
    return normalizePath(relative(sourceRoot, sourceFile))
  }
  return sourceFile
}

type ErrorRules = ErrorRule[]
export type ErrorRuleConfig = {
  version: string
  common?: {
    rules: ErrorRules
  }
  kotlin?: {
    rules: ErrorRules
  }
  swift?: {
    rules: ErrorRules
  }
  js?: {
    rules: ErrorRules
  }
  arkts?: {
    rules: ErrorRules
  }
}

export type ErrorRule = {
  platform?: string[]
  pattern: string
  flags?: string
  message: string
  suggestions?: string[]
}

/**
 * 解析错误信息
{
  "version": "1.0.0",
  "common": {
    "rules": [
      {
        "pattern": "Cannot find module '(.+)'",
        "flags": "i",
        "message": "模块错误：找不到模块 '$1'",
        "suggestions": ["检查模块路径", "确认模块已安装"]
      }
    ]
  },
  "js":{
    "rules": [
      {
        "platform": ["app-ios"],
        "pattern": "JS specific error (.+)",
        "flags": "i",
        "message": "JS 平台错误：$1",
        "suggestions": ["检查 JS 配置"]
      }
    ]
  },
  "kotlin": {
    "rules": [
      {
        "platform": ["app-android"],
        "pattern": "Android specific error (.+)",
        "flags": "i",
        "message": "Android 平台错误：$1",
        "suggestions": ["检查 Android 构建配置"]
      }
    ]
  },
  "swift": {
    "rules": [
      {
        "platform": ["app-ios"],
        "pattern": "iOS specific error (.+)",
        "flags": "i",
        "message": "iOS 平台错误：$1",
        "suggestions": ["检查 Xcode 配置"]
      }
    ]
  }
} 
 * @param error 错误信息
 * @returns 错误规则
 */
function normalizeErrorWithRule(error: string, rules: ErrorRule[]) {
  for (const rule of rules) {
    const re = new RegExp(rule.pattern, rule.flags)
    if (re.test(error)) {
      return error.replace(re, rule.message)
    }
  }
  return error
}

export function parseErrorWithRules(
  error: string,
  options: {
    language: 'kotlin' | 'swift' | 'js' | 'arkts'
    platform: 'app-android' | 'app-ios' | 'app-harmony' | 'mp-weixin' | 'web'
  }
) {
  try {
    const rules = getErrorRules(options.language, options.platform)
    if (rules.length === 0) {
      return error
    }
    return normalizeErrorWithRule(error, rules)
  } catch (e) {}
  return error
}

let errorRuleConfig: ErrorRuleConfig | null = null
function getErrorRuleConfig() {
  if (errorRuleConfig) {
    return errorRuleConfig
  }
  const filename = process.env.UNI_COMPILER_VALIDATION_RULES_PATH
  if (filename && fs.existsSync(filename)) {
    try {
      errorRuleConfig = JSON.parse(fs.readFileSync(filename, 'utf-8'))
    } catch (e) {}
  }
  if (!errorRuleConfig) {
    errorRuleConfig = {
      version: '1.0.0',
    }
  }
  return errorRuleConfig
}

function getErrorRules(
  language: 'kotlin' | 'swift' | 'js' | 'arkts',
  platform: 'app-android' | 'app-ios' | 'app-harmony' | 'mp-weixin' | 'web'
) {
  const errorRuleConfig = getErrorRuleConfig()
  const rules: ErrorRule[] = [...(errorRuleConfig.common?.rules || [])]
  if (errorRuleConfig[language]) {
    rules.push(...(errorRuleConfig[language]?.rules || []))
  }
  return rules.filter((rule) => {
    if (rule.platform && rule.platform.length > 0) {
      return rule.platform.includes(platform)
    }
    return true
  })
}
