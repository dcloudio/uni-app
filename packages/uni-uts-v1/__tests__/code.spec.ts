import { resolve } from 'path'
import { FORMATS, type GenProxyCodeOptions, genProxyCode } from '../src/code'
import { ERR_MSG_PLACEHOLDER } from '../src/utils'

const inputDir = resolve(__dirname, 'examples/uts')
const pluginDir = resolve(__dirname, 'examples/uts/utssdk/test-uts')
const uniModuleDir = resolve(__dirname, 'examples/uts/uni_modules/test-uts')
const uniModuleKeepAliveDir = resolve(
  __dirname,
  'examples/uts/uni_modules/test-keepAlive'
)
describe('code', () => {
  test('genProxyCode', async () => {
    const options: GenProxyCodeOptions = {
      moduleName: '测试',
      moduleType: 'built-in',
      id: 'test-uts',
      is_uni_modules: false,
      name: 'test-uts',
      namespace: 'uts.sdk.testUTS',
      extname: '.uts',
      androidComponents: { TestUTS: '' },
      inputDir,
    }
    const res = await genProxyCode(pluginDir, options)
    expect(res.replace(ERR_MSG_PLACEHOLDER, '')).toMatchSnapshot()
    expect(options.meta).toMatchSnapshot()
    expect(options.types).toMatchSnapshot()
  })
  test('genProxyCode cjs', async () => {
    expect(
      (
        await genProxyCode(pluginDir, {
          id: 'test-uts',
          is_uni_modules: false,
          name: 'test-uts',
          namespace: 'uts.sdk.testUTS',
          extname: '.uts',
          format: FORMATS.CJS,
          pluginRelativeDir: 'utssdk/test-uts',
          androidComponents: { TestUTS: '' },
          inputDir,
        })
      ).replace(ERR_MSG_PLACEHOLDER, '')
    ).toMatchSnapshot()
  })

  test('genProxyCode uni_modules', async () => {
    expect(
      (
        await genProxyCode(uniModuleDir, {
          id: 'test-uts',
          is_uni_modules: true,
          name: 'test-uts',
          namespace: 'uts.sdk.testUTS',
          extname: '.uts',
          androidComponents: {},
          inputDir,
        })
      ).replace(ERR_MSG_PLACEHOLDER, '')
    ).toMatchSnapshot()
  })

  test('genProxyCode uni_modules keepAlive', async () => {
    expect(
      (
        await genProxyCode(uniModuleKeepAliveDir, {
          id: 'test-keepAlive',
          is_uni_modules: true,
          name: 'test-keepAlive',
          namespace: 'uts.sdk.testKeepAlive',
          extname: '.uts',
          androidComponents: {},
          inputDir,
        })
      ).replace(ERR_MSG_PLACEHOLDER, '')
    ).toMatchSnapshot()
  })
})
