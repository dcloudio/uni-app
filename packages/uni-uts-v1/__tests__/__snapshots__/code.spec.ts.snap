// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code genProxyCode 1`] = `
"
const { registerUTSInterface, initUTSProxyClass, initUTSProxyFunction, initUTSPackageName, initUTSIndexClassName, initUTSClassName } = uni
const name = 'test-uts'
const moduleName = '测试'
const moduleType = 'built-in'
const errMsg = \`\`
const is_uni_modules = false
const pkg = /*#__PURE__*/ initUTSPackageName(name, is_uni_modules)
const cls = /*#__PURE__*/ initUTSIndexClassName(name, is_uni_modules)

export const TestUTSComponent = {}

registerUTSInterface('RequestTaskOptions',Object.assign({ moduleName, moduleType, errMsg, package: pkg, class: initUTSClassName(name, 'RequestTaskByJsProxy', is_uni_modules) }, {"methods":{"startByJs":{"async":false,"keepAlive":false,"params":[]},"abortByJs":{"async":false,"keepAlive":false,"params":[],"return":{"type":"interface","options":"RequestTaskOptions"}}},"props":[],"setters":{}} ))
export const onMemoryWarning = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'onMemoryWarningByJs', keepAlive: false, params: [{"name":"callback","type":"UTSCallback"}], return: ""})
export const offMemoryWarning = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'offMemoryWarningByJs', keepAlive: false, params: [{"name":"callback","type":"UTSCallback","default":"UTSNull"}], return: ""})
export default /*#__PURE__*/ initUTSProxyClass(Object.assign({ moduleName, moduleType, errMsg, package: pkg, class: initUTSClassName(name, 'UserByJs', is_uni_modules) }, {"constructor":{"params":[]},"methods":{"registerByJs":{"async":false,"keepAlive":false,"params":[]},"loginByJs":{"async":false,"keepAlive":false,"params":[]}},"staticMethods":{},"props":[],"staticProps":[],"setters":{},"staticSetters":{}} ))
export const a = 1
export const showToast1 = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'showToast1ByJs', keepAlive: false, params: [{"name":"msg","type":"string"}], return: ""})
export const showToast2 = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'showToast2ByJs', keepAlive: false, params: [{"name":"msg","type":"string"}], return: ""})
export const request = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'requestByJs', keepAlive: false, params: [], return: {"type":"interface","options":"RequestTaskOptions"}})
export const addUsers1 = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'addUsers1ByJs', keepAlive: false, params: [{"name":"users","type":"uts.sdk.testUTSUser1JSONObject"}], return: ""})
export const addUsers2 = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'addUsers2ByJs', keepAlive: false, params: [{"name":"users","type":"uts.sdk.testUTSUser1JSONObject"}], return: ""})
export const addUsers3 = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'addUsers3ByJs', keepAlive: false, params: [{"name":"user","type":"uts.sdk.testUTSUser1JSONObject"},{"name":"name","type":"string"}], return: ""})
export const addUsers4 = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'addUsers4ByJs', keepAlive: false, params: [{"name":"user","type":"uts.sdk.testUTSUser1JSONObject"},{"name":"name","type":"string"}], return: ""})
"
`;

exports[`code genProxyCode 2`] = `
{
  "components": [
    "TestUTS",
  ],
  "exports": {
    "RequestTask": {
      "type": "interface",
    },
    "User": {
      "type": "class",
    },
    "a": {
      "type": "var",
    },
    "addUsers1": {
      "params": [
        {
          "name": "users",
          "type": "uts.sdk.testUTSUser1JSONObject",
        },
      ],
      "type": "function",
    },
    "addUsers2": {
      "params": [
        {
          "name": "users",
          "type": "uts.sdk.testUTSUser1JSONObject",
        },
      ],
      "type": "function",
    },
    "addUsers3": {
      "params": [
        {
          "name": "user",
          "type": "uts.sdk.testUTSUser1JSONObject",
        },
        {
          "name": "name",
          "type": "string",
        },
      ],
      "type": "function",
    },
    "addUsers4": {
      "params": [
        {
          "name": "user",
          "type": "uts.sdk.testUTSUser1JSONObject",
        },
        {
          "name": "name",
          "type": "string",
        },
      ],
      "type": "function",
    },
    "offMemoryWarning": {
      "params": [
        {
          "default": "UTSNull",
          "name": "callback",
          "type": "UTSCallback",
        },
      ],
      "type": "function",
    },
    "onMemoryWarning": {
      "params": [
        {
          "name": "callback",
          "type": "UTSCallback",
        },
      ],
      "type": "function",
    },
    "request": {
      "params": [],
      "type": "function",
    },
    "showToast1": {
      "params": [
        {
          "name": "msg",
          "type": "string",
        },
      ],
      "type": "var",
    },
    "showToast2": {
      "params": [
        {
          "name": "msg",
          "type": "string",
        },
      ],
      "type": "var",
    },
  },
  "typeParams": [],
  "types": {
    "RequestTask": "interface",
    "ShowToast": "function",
    "TestUTSElement": "class",
    "uni": [],
  },
}
`;

exports[`code genProxyCode 3`] = `
{
  "alias": {},
  "class": {
    "User1": {
      "interfaces": [],
      "keepAliveMethods": [],
      "typeParams": false,
    },
  },
  "fn": {
    "ShowToast": [
      {
        "pat": {
          "optional": false,
          "span": {
            "ctxt": 0,
            "end": 37,
            "start": 26,
          },
          "type": "Identifier",
          "typeAnnotation": {
            "span": {
              "ctxt": 0,
              "end": 37,
              "start": 29,
            },
            "type": "TsTypeAnnotation",
            "typeAnnotation": {
              "kind": "string",
              "span": {
                "ctxt": 0,
                "end": 37,
                "start": 31,
              },
              "type": "TsKeywordType",
            },
          },
          "value": "msg",
        },
        "span": {},
        "type": "Parameter",
      },
    ],
  },
  "interface": {
    "RequestTask": {
      "decl": {
        "body": {
          "body": [
            {
              "computed": false,
              "isOverride": false,
              "key": {
                "optional": false,
                "span": {
                  "ctxt": 0,
                  "end": 85,
                  "start": 80,
                },
                "type": "Identifier",
                "value": "start",
              },
              "optional": false,
              "params": [],
              "readonly": false,
              "span": {
                "ctxt": 0,
                "end": 93,
                "start": 80,
              },
              "type": "TsMethodSignature",
              "typeAnn": {
                "span": {
                  "ctxt": 0,
                  "end": 93,
                  "start": 87,
                },
                "type": "TsTypeAnnotation",
                "typeAnnotation": {
                  "kind": "void",
                  "span": {
                    "ctxt": 0,
                    "end": 93,
                    "start": 89,
                  },
                  "type": "TsKeywordType",
                },
              },
              "typeParams": null,
            },
            {
              "computed": false,
              "isOverride": false,
              "key": {
                "optional": false,
                "span": {
                  "ctxt": 0,
                  "end": 101,
                  "start": 96,
                },
                "type": "Identifier",
                "value": "abort",
              },
              "optional": false,
              "params": [],
              "readonly": false,
              "span": {
                "ctxt": 0,
                "end": 116,
                "start": 96,
              },
              "type": "TsMethodSignature",
              "typeAnn": {
                "span": {
                  "ctxt": 0,
                  "end": 116,
                  "start": 103,
                },
                "type": "TsTypeAnnotation",
                "typeAnnotation": {
                  "span": {
                    "ctxt": 0,
                    "end": 116,
                    "start": 105,
                  },
                  "type": "TsTypeReference",
                  "typeName": {
                    "optional": false,
                    "span": {
                      "ctxt": 0,
                      "end": 116,
                      "start": 105,
                    },
                    "type": "Identifier",
                    "value": "RequestTask",
                  },
                  "typeParams": null,
                },
              },
              "typeParams": null,
            },
          ],
          "span": {
            "ctxt": 0,
            "end": 118,
            "start": 76,
          },
          "type": "TsInterfaceBody",
        },
        "declare": false,
        "extends": [],
        "id": {
          "optional": false,
          "span": {
            "ctxt": 0,
            "end": 75,
            "start": 64,
          },
          "type": "Identifier",
          "value": "RequestTask",
        },
        "span": {
          "ctxt": 0,
          "end": 118,
          "start": 54,
        },
        "type": "TsInterfaceDeclaration",
        "typeParams": null,
      },
      "returned": true,
    },
  },
  "uni": [],
}
`;

exports[`code genProxyCode cjs 1`] = `
"
const { registerUTSInterface, initUTSProxyClass, initUTSProxyFunction, initUTSPackageName, initUTSIndexClassName, initUTSClassName } = uni
const name = 'test-uts'
const moduleName = ''
const moduleType = ''
const errMsg = \`\`
const is_uni_modules = false
const pkg = /*#__PURE__*/ initUTSPackageName(name, is_uni_modules)
const cls = /*#__PURE__*/ initUTSIndexClassName(name, is_uni_modules)

const exports = { __esModule: true }

exports.TestUTSComponent = {}

registerUTSInterface('RequestTaskOptions',Object.assign({ moduleName, moduleType, errMsg, package: pkg, class: initUTSClassName(name, 'RequestTaskByJsProxy', is_uni_modules) }, {"methods":{"startByJs":{"async":false,"keepAlive":false,"params":[]},"abortByJs":{"async":false,"keepAlive":false,"params":[],"return":{"type":"interface","options":"RequestTaskOptions"}}},"props":[],"setters":{}} ))
exports.onMemoryWarning = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'onMemoryWarningByJs', keepAlive: false, params: [{"name":"callback","type":"UTSCallback"}], return: ""})
exports.offMemoryWarning = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'offMemoryWarningByJs', keepAlive: false, params: [{"name":"callback","type":"UTSCallback","default":"UTSNull"}], return: ""})
exports.default = initUTSProxyClass(Object.assign({ moduleName, moduleType, errMsg, package: pkg, class: initUTSClassName(name, 'UserByJs', is_uni_modules) }, {"constructor":{"params":[]},"methods":{"registerByJs":{"async":false,"keepAlive":false,"params":[]},"loginByJs":{"async":false,"keepAlive":false,"params":[]}},"staticMethods":{},"props":[],"staticProps":[],"setters":{},"staticSetters":{}} ))
const a = 1
exports.a = a
exports.showToast1 = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'showToast1ByJs', keepAlive: false, params: [{"name":"msg","type":"string"}], return: ""})
exports.showToast2 = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'showToast2ByJs', keepAlive: false, params: [{"name":"msg","type":"string"}], return: ""})
exports.request = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'requestByJs', keepAlive: false, params: [], return: {"type":"interface","options":"RequestTaskOptions"}})
exports.addUsers1 = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'addUsers1ByJs', keepAlive: false, params: [{"name":"users","type":"uts.sdk.testUTSUser1JSONObject"}], return: ""})
exports.addUsers2 = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'addUsers2ByJs', keepAlive: false, params: [{"name":"users","type":"uts.sdk.testUTSUser1JSONObject"}], return: ""})
exports.addUsers3 = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'addUsers3ByJs', keepAlive: false, params: [{"name":"user","type":"uts.sdk.testUTSUser1JSONObject"},{"name":"name","type":"string"}], return: ""})
exports.addUsers4 = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'addUsers4ByJs', keepAlive: false, params: [{"name":"user","type":"uts.sdk.testUTSUser1JSONObject"},{"name":"name","type":"string"}], return: ""})
uni.registerUTSPlugin('utssdk/test-uts', exports)
"
`;

exports[`code genProxyCode uni_modules 1`] = `
"
const { registerUTSInterface, initUTSProxyClass, initUTSProxyFunction, initUTSPackageName, initUTSIndexClassName, initUTSClassName } = uni
const name = 'test-uts'
const moduleName = ''
const moduleType = ''
const errMsg = \`\`
const is_uni_modules = true
const pkg = /*#__PURE__*/ initUTSPackageName(name, is_uni_modules)
const cls = /*#__PURE__*/ initUTSIndexClassName(name, is_uni_modules)



registerUTSInterface('ITestOptions',Object.assign({ moduleName, moduleType, errMsg, package: pkg, class: initUTSClassName(name, 'ITestByJsProxy', is_uni_modules) }, {"methods":{},"props":["name"],"setters":{"name":{"name":"name","type":"string"}}} ))
export const createTest = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'createTestByJs', keepAlive: false, params: [], return: {"type":"interface","options":"ITestOptions"}})
registerUTSInterface('A12Options',Object.assign({ moduleName, moduleType, errMsg, package: pkg, class: initUTSClassName(name, 'A12ByJsProxy', is_uni_modules) }, {"methods":{"test2ByJs":{"async":false,"keepAlive":false,"params":[]},"test1ByJs":{"async":false,"keepAlive":false,"params":[]},"test0ByJs":{"async":false,"keepAlive":false,"params":[]}},"props":[],"setters":{}} ))
export const testA3 = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'testA3ByJs', keepAlive: false, params: [], return: {"type":"interface","options":"A12Options"}})
"
`;

exports[`code genProxyCode uni_modules keepAlive 1`] = `
"
const { registerUTSInterface, initUTSProxyClass, initUTSProxyFunction, initUTSPackageName, initUTSIndexClassName, initUTSClassName } = uni
const name = 'test-keepAlive'
const moduleName = ''
const moduleType = ''
const errMsg = \`\`
const is_uni_modules = true
const pkg = /*#__PURE__*/ initUTSPackageName(name, is_uni_modules)
const cls = /*#__PURE__*/ initUTSIndexClassName(name, is_uni_modules)



registerUTSInterface('ITestKeepAliveOptions',Object.assign({ moduleName, moduleType, errMsg, package: pkg, class: initUTSClassName(name, 'ITestKeepAliveByJsProxy', is_uni_modules) }, {"methods":{"test1ByJs":{"async":false,"keepAlive":false,"params":[{"name":"callback","type":"UTSCallback"}]},"test2ByJs":{"async":false,"keepAlive":true,"params":[{"name":"callback","type":"UTSCallback"}]}},"props":[],"setters":{}} ))
export const test1 = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'test1ByJs', keepAlive: false, params: [{"name":"callback","type":"UTSCallback"}], return: ""})
export const test2 = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'test2ByJs', keepAlive: true, params: [{"name":"callback","type":"UTSCallback"}], return: ""})
export const Test = /*#__PURE__*/ initUTSProxyClass(Object.assign({ moduleName, moduleType, errMsg, package: pkg, class: initUTSClassName(name, 'TestByJs', is_uni_modules) }, {"constructor":{"params":[]},"methods":{"test1ByJs":{"async":false,"keepAlive":false,"params":[{"name":"callback","type":"UTSCallback"}]},"test2ByJs":{"async":false,"keepAlive":true,"params":[{"name":"callback","type":"UTSCallback"}]}},"staticMethods":{"test3ByJs":{"async":false,"keepAlive":false,"params":[{"name":"callback","type":"UTSCallback"}]},"test4ByJs":{"async":false,"keepAlive":true,"params":[{"name":"callback","type":"UTSCallback"}]}},"props":[],"staticProps":[],"setters":{},"staticSetters":{}} ))
export const createTest2 = /*#__PURE__*/ initUTSProxyFunction(false, { moduleName, moduleType, errMsg, main: true, package: pkg, class: cls, name: 'createTest2ByJs', keepAlive: false, params: [], return: {"type":"interface","options":"ITestKeepAliveOptions"}})
"
`;
